contrib/bsp/src/mill/contrib/BSP.scala

```diff
@@ -31,107 +31,126 @@ object BSP extends ExternalModule {
-    val executor = Executors.newCachedThreadPool()
+   * @return: mill.Command which executes the starting of the
+   *          server
+   */
+  def start(ev: Evaluator): Command[Unit] =
+    T.command {
+      val eval = new Evaluator(
+        ev.home,
+        ev.outPath,
+        ev.externalOutPath,
+        ev.rootModule,
+        ev.baseLogger,
+        ev.classLoaderSig,
+        ev.workerCache,
+        ev.env,
+        false
+      )
+      val millServer = new MillBuildServer(eval, bspProtocolVersion, BuildInfo.millVersion)
+      val executor = Executors.newCachedThreadPool()
 
-    val stdin = System.in
-    val stdout = System.out
-    try {
-      val launcher = new Launcher.Builder[BuildClient]()
-        .setOutput(stdout)
-        .setInput(stdin)
-        .setLocalService(millServer)
-        .setRemoteInterface(classOf[BuildClient]).
-        traceMessages(new PrintWriter((os.pwd / "bsp.log").toIO))
-        .setExecutorService(executor)
-        .create()
-      millServer.onConnectWithClient(launcher.getRemoteProxy)
-      val listening = launcher.startListening()
-      millServer.cancelator = () => listening.cancel(true)
-      val voidFuture = listening.get()
-    } catch {
-      case _: CancellationException => System.err.println("The mill server was shut down.")
-      case e: Exception =>
-        System.err.println("An exception occured while connecting to the client.")
-        System.err.println("Cause: " + e.getCause)
-        System.err.println("Message: " + e.getMessage)
-        System.err.println("Exception class: " + e.getClass)
-        System.err.println("Stack Trace: " + e.getStackTrace)
-    } finally {
-      System.err.println("Shutting down executor")
-      executor.shutdown()
+      val stdin = System.in
+      val stdout = System.out
+      try {
+        val launcher = new Launcher.Builder[BuildClient]()
+          .setOutput(stdout)
+          .setInput(stdin)
+          .setLocalService(millServer)
+          .setRemoteInterface(classOf[BuildClient])
+          .traceMessages(new PrintWriter((os.pwd / "bsp.log").toIO))
+          .setExecutorService(executor)
+          .create()
+        millServer.onConnectWithClient(launcher.getRemoteProxy)
+        val listening = launcher.startListening()
+        millServer.cancelator = () => listening.cancel(true)
+        listening.get()
+        ()
+      } catch {
+        case _: CancellationException => System.err.println("The mill server was shut down.")
+        case e: Exception =>
+          System.err.println(
+            s"""An exception occured while connecting to the client.
+              |Cause: ${e.getCause}
+              |Message: ${e.getMessage}
+              |Exception class: ${e.getClass}
+              |Stack Trace: ${e.getStackTrace}""".stripMargin
+          )
+      } finally {
+        System.err.println("Shutting down executor")
```

Comment:
Shouldn't we use `T.log.error` here?

Reply:
Good point. I will do the change in a bit.